crates/sdk/macros/src/logic/method.rs

```diff
@@ -19,6 +20,51 @@ pub struct PublicLogicMethod<'a> {
     has_refs: bool,
 }
 
+pub struct InitMethod<'a> {
+    self_: syn::Path,
+    name: &'a syn::Ident,
+    args: Vec<arg::LogicArgTyped<'a>>,
+}
+
+impl<'a> ToTokens for LogicMethod<'a> {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        match self {
+            LogicMethod::Public(method) => method.to_tokens(tokens),
+            LogicMethod::Private => {}
+            LogicMethod::Init(method) => method.to_tokens(tokens),
+        }
+    }
+}
+
+impl<'a> ToTokens for InitMethod<'a> {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let self_ = &self.self_;
+        let name = &self.name;
+        let args = &self.args;
+
+        let arg_idents = args.iter().map(|arg| &*arg.ident).collect::<Vec<_>>();
+        let arg_types = args.iter().map(|arg| &arg.ty).collect::<Vec<_>>();
+
+        quote! {
+            #[cfg(target_arch = "wasm32")]
+            #[no_mangle]
+            pub extern "C" fn init() {
+                ::calimero_sdk::env::setup_panic_hook();
+
+                ::calimero_sdk::event::register::<#self_>();
+
+                let input = ::calimero_sdk::env::input().expect("Init parameters are required");
+                let (#(#arg_idents),*): (#(#arg_types),*) = ::calimero_sdk::serde_json::from_slice(&input)
```

Comment:
For posterity, the reason we use struct definitions is so you can provide a JSON payload with the field names in an object instead of just an array of parameters.

Worth noting: `serde` also attempts to deserialize an array of parameters, so we get the best of both worlds.